plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php

```diff
@@ -347,8 +355,88 @@ private function handle_data_sync_option_changed( string $feature_id ) {
 		// some reason it was ongoing while it was disabled).
 		if ( $data_sync_is_enabled ) {
 			$this->batch_processing_controller->enqueue_processor( DataSynchronizer::class );
+			$this->maybe_schedule_pending_sync_check();
 		} else {
 			$this->batch_processing_controller->remove_processor( DataSynchronizer::class );
+			$this->maybe_unschedule_pending_sync_check();
+		}
+	}
+
+	/**
+	 * Schedule a recurring event to check for pending order syncs.
+	 *
+	 * Wrapper to avoid throwing an error if Action Scheduler isn't initialized yet.
+	 *
+	 * @return void
+	 */
+	private function maybe_schedule_pending_sync_check() {
+		if ( ! ActionScheduler::is_initialized() ) {
+			return;
+		}
+
+		$has_scheduled_action = WC()->queue()->search( array(
+			'hook'   => self::SYNC_CHECK_EVENT_HOOK,
+			'status' => array( 'in-progress', 'pending' ),
+		) );
+
+		if ( ! $has_scheduled_action ) {
+			WC()->queue()->schedule_recurring(
+				time() + HOUR_IN_SECONDS,
+				6 * HOUR_IN_SECONDS,
+				self::SYNC_CHECK_EVENT_HOOK
+			);
+		}
+	}
+
+	/**
+	 * Remove any previously scheduled events to check for pending order syncs.
+	 *
+	 * Wrapper to avoid throwing an error if Action Scheduler isn't initialized yet.
+	 *
+	 * @return void
+	 */
+	private function maybe_unschedule_pending_sync_check() {
+		if ( ! ActionScheduler::is_initialized() ) {
+			return;
+		}
+
+		WC()->queue()->cancel_all( self::SYNC_CHECK_EVENT_HOOK );
+	}
+
+	/**
+	 * When sync is turned on, check to see if there are any pending syncs and if so, queue the processor.
+	 *
+	 * Runs as a recurring callback for SYNC_CHECK_EVENT_HOOK.
+	 *
+	 * @return void
+	 */
+	private function check_for_pending_syncs() {
+		if ( ! $this->data_synchronizer->data_sync_is_enabled() ) {
```

Comment:
🤔 But if they have data sync disabled, the event to run this callback will be unscheduled, so it wouldn't run anyway. I think you're on to something that it would be interesting to allow for an... async sync option so there isn't a real-time performance hit. But I think it would take a bit more to implement, and seems like would be better addressed in a separate PR.

Reply:
> async sync option so there isn't a real-time performance hit.

seems like we need this to support high-volume sites (such as wccom) as an intermediatory steps in turning off the sync. knowing this, would you change anything in this PR? Perhaps we can use the similar mechanism for both these things? wdyt?

Reply:
If we go this route, it seems like we'll need to evolve the data sync setting away from a simple checkbox to something else. Or, maybe it's still a checkbox, but when it's checked, more options are revealed and you then have to decide between "real-time" syncing and periodic sync?

If we can figure out the UI for this, I think this PR gets us most of the way to the implementation. Any other thoughts you have about the UI?